store/neo4j.go

```diff
@@ -0,0 +1,201 @@
+package store
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/ONSdigital/dp-code-list-api/datastore"
+	"github.com/ONSdigital/dp-code-list-api/models"
+	"github.com/ONSdigital/go-ns/log"
+	bolt "github.com/johnnadratowski/golang-neo4j-bolt-driver"
+	"github.com/johnnadratowski/golang-neo4j-bolt-driver/structures/graph"
+	"github.com/pkg/errors"
+)
+
+const (
+	getCodeListsQuery = "MATCH (i) WHERE i:_%s%s RETURN distinct labels(i) as labels, i"
+	getCodeListQuery  = "MATCH (i:_%s:`_name_%s`) RETURN i"
+)
+
+// NeoDataStore represents the necessary information to access
+// neo4j
+type NeoDataStore struct {
+	pool          DBPool
+	codeListLabel string
+}
+
+// DBPool contains the methods to control access to the Neo4J
+// database pool
+type DBPool interface {
+	OpenPool() (bolt.Conn, error)
+	Close() error
+}
+
+// Close is a wrapper for the neo pool close
+func (n *NeoDataStore) Close() error {
+	return n.pool.Close()
+}
+
+// CreateNeoDataStore allows the creation of a NeoDataStore
+func CreateNeoDataStore(addr, codelistLabel string, conns int) (n *NeoDataStore, err error) {
+	store, err := bolt.NewClosableDriverPool(addr, conns)
+	if err != nil {
+		return
+	}
+
+	n = &NeoDataStore{
+		pool:          store,
+		codeListLabel: codelistLabel,
+	}
+
+	return
+}
+
+// GetCodeLists returns a list of code lists
+func (n *NeoDataStore) GetCodeLists(ctx context.Context, filterBy string) (*models.CodeListResults, error) {
+	logData := log.Data{}
+	if len(filterBy) > 0 {
+		logData = log.Data{"filter_by": filterBy}
+	}
+	log.InfoCtx(ctx, "about to query neo4j for code lists", logData)
+
+	conn, err := n.pool.OpenPool()
+	if err != nil {
+		return nil, errors.WithMessage(err, "could not retrieve connection from pool")
+	}
+
+	defer func() {
+		if err := conn.Close(); err != nil {
+			log.ErrorCtx(ctx, errors.WithMessage(err, "could not close neo4j connection"), nil)
+		}
+	}()
+
+	if len(filterBy) > 0 {
+		filterBy = ":_" + filterBy
+	}
+
+	query := fmt.Sprintf(getCodeListsQuery, n.codeListLabel, filterBy)
+
+	rows, err := conn.QueryNeo(query, nil)
+	if err != nil {
+		return nil, errors.WithMessage(err, "could not run neo4j query")
+	}
+
+	var row []interface{}
+
+	codeLists := &models.CodeListResults{}
+	for row, _, err = rows.NextNeo(); err == nil; row, _, err = rows.NextNeo() {
+		if len(row) < 2 {
+			return nil, errors.Errorf("expected at least two rows, got %d", len(row))
+		}
+		props := row[1].(graph.Node).Properties
+		labels := row[1].(graph.Node).Labels
+
+		name := props["label"].(string)
+
+		var label string
+		for _, l := range labels {
+			if strings.Contains(l, "_name_") {
+				label = l
+				break
+			}
+		}
+
+		neoLabel := strings.Replace(label, `_name_`, "", -1)
```

Comment:
couldnt this happen within the for loop, and then you wouldnt need both neoLabel and label, as label doesnt appear to be used beyond this point?